c++/src/capnp/compat/json.c++

```diff
@@ -372,6 +382,251 @@ Orphan<DynamicValue> JsonCodec::decode(
 
 // -----------------------------------------------------------------------------
 
+namespace _ {  // private
+
+class Parser {
+public:
+  Parser(kj::ArrayPtr<const char> input) : input_(input), remaining_(input_) {}
+  void parseValue(JsonValue::Builder &output) {
+    consumeWhitespace();
+    KJ_REQUIRE(remaining_.size() > 0, "JSON message ends prematurely.");
+
+    switch (nextChar()) {
+      case 'n': consume(NULL_); output.setNull();         break;
+      case 'f': consume(FALSE); output.setBoolean(false); break;
+      case 't': consume(TRUE);  output.setBoolean(true);  break;
+      case '"': parseString(output); break;
+      case '[': parseArray(output);  break;
+      case '{': parseObject(output); break;
+      // TODO(security): We could check for numbers more carefully instead of
+      // relying on strtod.
+      default: parseNumber(output); break;
+    }
+  }
+
+  void parseNumber(JsonValue::Builder &output) {
+    // TODO(someday): strtod allows leading +, while JSON grammar does not.
+    // strtod consumes leading whitespace, so we don't have to.
+    char *numEnd;
+    output.setNumber(std::strtod(remaining_.begin(), &numEnd));
```

Comment:
Also I think `strtod()` will accept redundant leading zeros. The JSON spec disallows these presumably due to ambiguity with octal. It's probably best that we also disallow them, rather than possibly misinterpret them.